ffc/codegeneration/definitions.py

```diff
@@ -16,34 +16,18 @@
 def num_coordinate_component_dofs(coordinate_element):
     """Get the number of dofs for a coordinate component for this degree.
 
-    This is a local hack that works for Lagrange 1-3, better
-    would be to get this passed by ffc from fiat through the ir.
-    The table data is to messy to figure out a clean design for that quickly.
     """
-    from ufl.cell import num_cell_entities
-    degree = coordinate_element.degree()
-    cell = coordinate_element.cell()
-    tdim = cell.topological_dimension()
-    cellname = cell.cellname()
+    from ffc.fiatinterface import create_element
+    fiat_elements = create_element(coordinate_element).elements()
+    # Extracting only first component degrees of freedom from FIAT
+    fiat_element = fiat_elements[0]
+    assert(all(isinstance(element, type(fiat_element)) for element in fiat_elements))
+
+    dof_dict = fiat_element.entity_dofs()
     d = 0
-    for entity_dim in range(tdim + 1):
-        # n = dofs per cell entity
-        if entity_dim == 0:
-            n = 1
-        elif entity_dim == 1:
-            n = degree - 1
-        elif entity_dim == 2:
-            if cellname == "quadrilateral":
-                n = (degree - 1) * (degree - 1)
-            else:
-                n = (degree - 2) * (degree - 1) // 2
-        elif entity_dim == 3:
-            n = (degree - 3) * (degree - 2) * (degree - 1) // 6
-        else:
-            raise RuntimeError("Entity dimension out of range")
-        # Accumulate
-        num_entities = num_cell_entities[cellname][entity_dim]
-        d += num_entities * n
+    for entity_dim in dof_dict.keys():
```

Comment:
What about using
```
    def space_dimension(self):
        """Return the dimension of the finite element space."""
        return len(self.dual_basis())
```
?

`entity_dofs` doesnt return e.g. dofs of DG elements, which live in the interior of the cell, you'd need to use `entity_closure_dofs`. I know coordinate map is not discontinuous, but just for clarity.

Reply:
Sure, seems ok. Reason for the other version was to extend the old hack to arbitrary lagrangian elements.

Reply:
One more comment. You can notice currently, that FIAT is interfaced only in `fiatiinterface.py`. Can you refactor a but and query FIAT library only in `fiatinterface.py`?

Reply:
Is that true? I can see it accessed in: codegeneration/access.py

Reply:
Ah, sorry. I thought you were importing FIAT library, its fine like this, nvm.